src/fluid.rs

```diff
@@ -56,128 +65,131 @@ pub struct Fluid {
-
-    fn coordinate_to_idx(x: u32, y: u32, size: &u32) -> usize {
-        let x = &Fluid::constrain(x, 0, size - 1);
-        let y = &Fluid::constrain(y, 0, size - 1);
+    fn coordinate_to_idx(x: u32, y: u32, size: u32) -> usize {
+        let x = x.clamp(0, size - 1);
+        let y = y.clamp(0, size - 1);
         (x + (y * size)) as usize
     }
 
     fn add_density(&mut self, x: u32, y: u32, amount: f32) {
-        let idx: usize = Fluid::coordinate_to_idx(x, y, &self.simulation_configs.size);
+        let idx: usize = Fluid::coordinate_to_idx(x, y, self.simulation_configs.size);
         self.density[idx] += amount;
         self.s[idx] += amount;
     }
 
     fn add_velocity(&mut self, x: u32, y: u32, amount_x: f32, amount_y: f32) {
-        let idx: usize = Fluid::coordinate_to_idx(x, y, &self.simulation_configs.size);
+        let idx: usize = Fluid::coordinate_to_idx(x, y, self.simulation_configs.size);
         self.velocities_x[idx] += amount_x;
         self.velocities_y[idx] += amount_y;
     }
 
     fn render_density(&self, imgbuf: &mut image::RgbImage, frame_number: u32) {
         for (x, y, pixel) in imgbuf.enumerate_pixels_mut() {
             let density =
-                self.density[Fluid::coordinate_to_idx(x, y, &self.simulation_configs.size)];
+                self.density[Fluid::coordinate_to_idx(x, y, self.simulation_configs.size)];
             *pixel = image::Rgb([(density * 255.0) as u8, 200, density as u8]);
         }
         let img_name = format!("rendered_images/density{}.jpg", frame_number);
         imgbuf.save(img_name).unwrap();
     }
 
-    fn set_boundaries(b: u32, x: &mut [f32], size: &u32) {
+    /// Process the edge walls, a.k.a. turn the velocities in the opposite way
+    /// The function is used not only for velocities, but for density, etc, where
+    /// A turn would have no meaning at all or will cause logical errors.
+    /// Use ContainerWall::DefaultWall if no turn is needed
+    fn set_boundaries(edge_wall: ContainerWall, x: &mut [f32], size: u32) {
         for i in 1..size - 1 {
-            x[Fluid::coordinate_to_idx(i, 0, size)] = if b == 2 {
+            x[Fluid::coordinate_to_idx(i, 0, size)] = if edge_wall == ContainerWall::East {
                 -x[Fluid::coordinate_to_idx(i, 1, size)]
             } else {
                 x[Fluid::coordinate_to_idx(i, 1, size)]
             };
-            x[Fluid::coordinate_to_idx(i, size - 1, size)] = if b == 2 {
+            x[Fluid::coordinate_to_idx(i, size - 1, size)] = if edge_wall == ContainerWall::East {
                 -x[Fluid::coordinate_to_idx(i, size - 2, size)]
             } else {
                 x[Fluid::coordinate_to_idx(i, size - 2, size)]
             };
         }
 
         for j in 1..size - 1 {
-            x[Fluid::coordinate_to_idx(0, j, size)] = if b == 1 {
-                -x[Fluid::coordinate_to_idx(1, j, size)]
+            x[Fluid::coordinate_to_idx(0, j, *&size)] = if edge_wall == ContainerWall::North {
```

Comment:
why `*&size`, this looks wrong. Should be just `size`

Reply:
Fixed